braindecode/models/ifnet.py

```diff
@@ -0,0 +1,419 @@
+        frequency convolutional neural network for enhancing motor imagery
+        decoding from EEG.
+        https://github.com/Jiaheng-Wang/IFNet
+    """
+
+    def __init__(
+        self,
+        # Braindecode parameters
+        n_chans=None,
+        n_outputs=None,
+        n_times=None,
+        chs_info=None,
+        input_window_seconds=None,
+        sfreq=None,
+        # Model-specific parameters
+        bands: list[tuple[float, float]] | int | None = [(4.0, 16.0), (16, 40)],
+        n_filters_spat: int = 64,
+        kernel_sizes: tuple[int, int] = (63, 31),
+        stride_factor: int = 8,
+        drop_prob: float = 0.5,
+        activation: nn.Module = nn.GELU,
+        verbose: bool = False,
+        filter_parameters: dict = {},
+    ):
+        super().__init__(
+            n_chans=n_chans,
+            n_outputs=n_outputs,
+            chs_info=chs_info,
+            n_times=n_times,
+            input_window_seconds=input_window_seconds,
+            sfreq=sfreq,
+        )
+        del n_outputs, n_chans, chs_info, n_times, input_window_seconds, sfreq
+
+        self.bands = bands
+        self.n_filters_spat = n_filters_spat
+        self.stride_factor = stride_factor
+        self.kernel_sizes = kernel_sizes
+        self.verbose = verbose
+        self.filter_parameters = filter_parameters
+        self.drop_prob = drop_prob
+        self.activation = activation
+        self.filter_parameters = filter_parameters
+
+        # Layers
+        # Following paper nomenclature
+        self.spectral_filtering = FilterBankLayer(
+            n_chans=self.n_chans,
+            sfreq=self.sfreq,
+            band_filters=self.bands,
+            verbose=verbose,
+            **self.filter_parameters,
+        )
+        # As we have an internal process to create the bands,
+        # we get the values from the filterbank
+        self.n_bands = self.spectral_filtering.n_bands
+
+        # My interpretation from the TABLE VII IFNet Architecture from the
+        # paper.
+        self.ensuredim = Rearrange(
+            "batch nbands chans time -> batch (nbands chans) time"
+        )
+
+        # SpatioTemporal Feature Block
+        self.feature_block = _SpatioTemporalFeatureBlock(
+            in_channels=self.n_chans * self.spectral_filtering.n_bands,
+            out_channels=self.n_filters_spat,
+            kernel_sizes=self.kernel_sizes,
+            stride_factor=self.stride_factor,
+            n_bands=self.spectral_filtering.n_bands,
+            drop_prob=self.drop_prob,
+            activation=self.activation,
+            n_times=self.n_times,
+        )
+
+        # Final classification layer
+        self.final_layer = nn.LazyLinear(
+            # LinearWithConstraint(
+            # in_features=self.n_filters_spat * (self.n_times // self.patch_size),
+            out_features=self.n_outputs,
+            # max_norm=0.5,
+        )
```

Comment:
Why are you not using `LinearWithConstraint` [like they do](https://github.com/Jiaheng-Wang/IFNet/blob/00ca53e29d2318a4cf7259878a4e42a7e00c6211/TEEGM/models/IFNetV2.py#L138)?

Reply:
my bad